src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithReferencedProductsIT.java

```diff
@@ -0,0 +1,470 @@
+    }
+
+    private void clearSyncTestCollections() {
+        errorCallBackMessages = new ArrayList<>();
+        errorCallBackExceptions = new ArrayList<>();
+        warningCallBackMessages = new ArrayList<>();
+    }
+
+    private ProductSyncOptions buildSyncOptions() {
+        final Consumer<String> warningCallBack = warningMessage -> warningCallBackMessages.add(warningMessage);
+
+        return ProductSyncOptionsBuilder.of(CTP_TARGET_CLIENT)
+                                        .errorCallback(this::collectErrors)
+                                        .warningCallback(warningCallBack)
+                                        .build();
+    }
+
+    private void collectErrors(final String errorMessage, final Throwable exception) {
+        errorCallBackMessages.add(errorMessage);
+        errorCallBackExceptions.add(exception);
+    }
+
+    @AfterAll
+    static void tearDown() {
+        deleteProductSyncTestData(CTP_TARGET_CLIENT);
+    }
+
+    @Test
+    void sync_withProductReferenceAsAttribute_shouldCreateProductReferencingExistingProduct() {
+        // preparation
+        final AttributeDraft productReferenceAttribute =
+            AttributeDraft.of("product-reference", Reference.of(Product.referenceTypeId(), product.getKey()));
+        final ProductVariantDraft masterVariant = ProductVariantDraftBuilder
+            .of()
+            .sku("sku")
+            .key("new-product-master-variant")
+            .attributes(productReferenceAttribute)
+            .build();
+
+        final ProductDraft productDraftWithProductReference = ProductDraftBuilder
+            .of(productType, ofEnglish("productName"), ofEnglish("productSlug"), masterVariant)
+            .key("new-product")
+            .build();
+
+
+        // test
+        final ProductSync productSync = new ProductSync(syncOptions);
+        final ProductSyncStatistics syncStatistics =
+            productSync
+                .sync(singletonList(productDraftWithProductReference))
+                .toCompletableFuture()
+                .join();
+
+        // assertion
+        assertThat(syncStatistics).hasValues(1, 1, 0, 0);
+        assertThat(errorCallBackExceptions).isEmpty();
+        assertThat(errorCallBackMessages).isEmpty();
+        assertThat(warningCallBackMessages).isEmpty();
+
+
+        final Product createdProduct = CTP_TARGET_CLIENT
+            .execute(ProductByKeyGet.of(productDraftWithProductReference.getKey()))
+            .toCompletableFuture()
+            .join();
+
+        final Optional<Attribute> createdProductReferenceAttribute =
+            createdProduct.getMasterData().getStaged().getMasterVariant()
+                          .findAttribute(productReferenceAttribute.getName());
+
+        assertThat(createdProductReferenceAttribute).hasValueSatisfying(attribute -> {
+            assertThat(attribute.getValueAsJsonNode().get("typeId").asText()).isEqualTo("product");
```

Comment:
try to use constants for id and typeId magic strings

Reply:
7d8b250

636205c